### PR TITLE
701 - Fixes #840: removes manual reference counting (#841)

### LINKED ISSUE #840: Remove manual reference counting from ServiceRegistrationBasePrivate and ServiceReferenceBasePrivate
Currently, destruction of ServiceRegistrationBasePrivate and ServiceReferenceBasePrivate occurs when manually tracked reference counters fall to 0. 

This should be updated to use built in types, like std::shared_ptr

### PR DESCRIPTION
**Pull request by insi-eb**
Cherry-picked 062fd47b2d91cc1764393b4b4bf412b421f4d4c4 / #841 with adaptations for C++14:
* move C++17-only initialization statements out of if clauses (ServiceReferenceBasePrivate.cpp)

Signed-off-by: Ingmar Sittl <ingmar.sittl@elektrobit.com>

### CODE REVIEW

File: framework/src/service/ServiceReferenceBasePrivate.cpp [hunk 1]
```diff
@@ -82,14 +84,14 @@ namespace cppmicroservices
...
+                auto l = coreInfo->properties.Lock();
                 US_UNUSED(l);
                 for (auto const& clazz : ref_any_cast<std::vector<std::string>>(
-                         registration->properties.ValueByRef_unlocked(Constants::OBJECTCLASS)))
+                         coreInfo->properties.ValueByRef_unlocked(Constants::OBJECTCLASS)))
                 {
                     if (smap->find(clazz) == smap->end() && clazz != "org.cppmicroservices.factory")
                     {
-                        if (auto bundle_ = registration->bundle.lock())
+                        if (auto bundle_ = coreInfo->bundle_.lock())
```
Comment (jeffdiclemente) [inline framework/src/service/ServiceReferenceBasePrivate.cpp:94]: the initialization statements are still in. is this intended?

**Review by jeffdiclemente — COMMENTED** [review]
@insi-eb I noticed there are still multiple initialization if statements in `ServiceReferenceBasePrivate.cpp`. I'm flagging it to make sure you're aware. let me know if you want to make additional changes or merge this PR as is. Thanks.

File: framework/src/service/ServiceReferenceBasePrivate.cpp (hunk 1 above)
Comment (insi-eb) [reply framework/src/service/ServiceReferenceBasePrivate.cpp:94]: This actually is a bool-convertible condition expression, not the C++17-only `if (init-statement ; condition)`. I also was not consciously aware that this is legal C++14, but for these PRs I only fix things the compilers complains about in C++14-mode.

TLDR; all is fine, this is legal C++14
